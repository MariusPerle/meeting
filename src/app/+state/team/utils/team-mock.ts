import { TeamMember } from './team.interface';

export function teamMemberFactory(
    overwrite: Pick<TeamMember, 'id'> & Partial<TeamMember>,
): TeamMember {
    return {
        name: `Member`,
        ...overwrite,
    };
}
