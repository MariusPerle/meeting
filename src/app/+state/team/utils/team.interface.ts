export interface TeamMemberWithRandomPosition extends TeamMember {
    randomPosition: number;
}

export interface TeamMember {
    id: string;
    name: string;
}

export interface RandomMember {
    id: string;
}

export function isTeamMember(data: object): data is TeamMember {
    return (
        'id' in data &&
        typeof data.id === 'string' &&
        'name' in data &&
        typeof data.name === 'string'
    );
}
