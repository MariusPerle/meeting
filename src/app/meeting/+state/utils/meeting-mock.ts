import { Record, RecordMember } from './meeting.interface';

export function recordFactory(
    overwrite: Pick<Record, 'id'> & Partial<Record>,
): Record {
    return {
        name: `Meeting`,
        date: new Date('2022-01-01'),
        members: [],
        ...overwrite,
    };
}

export function recordMemberFactory(
    overwrite: Pick<RecordMember, 'id'> & Partial<RecordMember>,
): RecordMember {
    return {
        time: 0,
        note: '',
        attended: true,
        participated: false,
        ...overwrite,
    };
}
