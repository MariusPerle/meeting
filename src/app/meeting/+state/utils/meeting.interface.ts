export interface Record {
    id: string;
    name: string;
    date: Date;
    members: RecordMember[];
}

export interface RecordMember {
    id: string;
    time: number;
    note: string;
    attended: boolean;
    participated: boolean;
}

export function isMeeting(data: object): data is Record {
    return (
        'id' in data &&
        typeof data.id === 'string' &&
        'name' in data &&
        typeof data.name === 'string' &&
        'date' in data &&
        data.date instanceof Date &&
        !isNaN(data.date.getTime()) &&
        'members' in data &&
        Array.isArray(data.members) &&
        data.members.every(isMeetingMember)
    );
}

export function isMeetingMember(data: object): data is RecordMember {
    return (
        'id' in data &&
        typeof data.id === 'string' &&
        'time' in data &&
        typeof data.time === 'number' &&
        'note' in data &&
        typeof data.note === 'string' &&
        'attended' in data &&
        typeof data.attended === 'boolean' &&
        'participated' in data &&
        typeof data.participated === 'boolean'
    );
}
