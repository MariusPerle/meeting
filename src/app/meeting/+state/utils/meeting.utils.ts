import { isMeeting, Record } from './meeting.interface';

export const meetingsStoreKey = 'meetings';

export function loadMeetingsFromLocalStorage(): Record[] {
    const localStorageValue = localStorage.getItem(meetingsStoreKey);
    if (!localStorageValue) {
        return [];
    }
    const meetings = JSON.parse(localStorageValue).map((meeting: Record) => ({
        ...meeting,
        date: new Date(meeting.date),
    }));

    if (!(Array.isArray(meetings) && meetings.every(isMeeting))) {
        return [];
    }
    return meetings;
}

export function saveToLocalStorage(history: Record[]) {
    localStorage.setItem(meetingsStoreKey, JSON.stringify(history));
}
