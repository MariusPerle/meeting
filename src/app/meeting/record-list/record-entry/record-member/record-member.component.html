<tr [ngClass]="{even: even()}">
  <th scope="row">{{ name$ | async }}</th>
  <td>{{ recordMember().note }}</td>
  <td class="status">
    @if (recordMember().time) {
      {{ recordMember().time | date: 'mm:ss' }}
    } @else {
      {{ status }}
    }
  </td>
  <td class="styling-column"></td>
</tr>
