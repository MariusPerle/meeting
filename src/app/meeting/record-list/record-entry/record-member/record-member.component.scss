:host {
    tr {
        border-bottom: 1px solid var(--bs-table-border-color);


        &.even {
            background: rgba(242,242,242,0.77);
        }
        td,th {
            padding: 0.5rem 0.5rem;
        }

        .status {
            text-align: end;
        }

        .styling-column {
            width: 0;
        }
    }
}