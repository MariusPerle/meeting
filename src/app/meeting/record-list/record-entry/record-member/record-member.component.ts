import { Component, inject, input, OnInit } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RecordMember } from '../../../+state/utils/meeting.interface';
import { Store } from '@ngrx/store';
import { MeetingStore } from '../../../+state/meeting.reducer';
import { selectMember } from '../../../../+state/team/team.selectors';
import { map, Observable } from 'rxjs';

@Component({
    selector: 'daily-record-member',
    imports: [CommonModule],
    templateUrl: './record-member.component.html',
    styleUrl: './record-member.component.scss',
})
export class RecordMemberComponent implements OnInit {
    recordMember = input.required<RecordMember>();
    even = input.required<boolean>();
    private readonly store: Store<MeetingStore> = inject(Store);

    name$!: Observable<string>;

    get status() {
        return !this.recordMember().attended
            ? 'absent'
            : this.recordMember().participated
              ? 'participated'
              : 'attended';
    }

    ngOnInit(): void {
        this.name$ = this.store
            .select(selectMember(this.recordMember().id))
            .pipe(map((member) => member?.name ?? 'no name'));
    }
}
