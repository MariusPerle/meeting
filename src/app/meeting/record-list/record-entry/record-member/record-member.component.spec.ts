import { ComponentFixture, TestBed } from '@angular/core/testing';
import { RecordMemberComponent } from './record-member.component';
import { provideMockStore } from '@ngrx/store/testing';
import { teamMemberFactory } from '../../../../+state/team/utils/team-mock';
import { Store } from '@ngrx/store';
import { of } from 'rxjs';
import { recordMemberFactory } from '../../../+state/utils/meeting-mock';

describe('RecordMemberComponent', () => {
    let component: RecordMemberComponent;
    let fixture: ComponentFixture<RecordMemberComponent>;
    const memberId = '1';

    beforeEach(async () => {
        await TestBed.configureTestingModule({
            imports: [RecordMemberComponent],
            providers: [provideMockStore()],
        }).compileComponents();

        fixture = TestBed.createComponent(RecordMemberComponent);
        component = fixture.componentInstance;
        fixture.componentRef.setInput(
            'recordMember',
            recordMemberFactory({
                id: memberId,
                note: 'note',
                time: 1000,
            }),
        );
        fixture.componentRef.setInput('even', true);

        // mock store
        jest.spyOn(
            (component as unknown as { store: Store }).store,
            'select',
        ).mockReturnValue(of(teamMemberFactory({ id: memberId })));
        fixture.detectChanges();
    });

    it('should display member with time', () => {
        expect(fixture).toMatchSnapshot();
    });

    it('should display member status if no time was recorded', () => {
        fixture.componentRef.setInput(
            'recordMember',
            recordMemberFactory({
                id: memberId,
                time: 0,
            }),
        );
        fixture.detectChanges();
        expect(
            fixture.nativeElement.querySelector('.status').textContent,
        ).toContain('attended');
    });
});
