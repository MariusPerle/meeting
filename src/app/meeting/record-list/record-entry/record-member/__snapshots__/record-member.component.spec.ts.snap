// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecordMemberComponent should display member with time 1`] = `
<meeting-record-member
  even={[Function Function]}
  name$={[Function Observable]}
  recordMember={[Function Function]}
  store={[Function _MockStore]}
>
  <tr
    class="even"
  >
    <th
      scope="row"
    >
      Member
    </th>
    <td>
      note
    </td>
    <td
      class="status"
    >
       00:01 
    </td>
    <td
      class="styling-column"
    />
  </tr>
</meeting-record-member>
`;
